src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader_MultiSegment.cs

```diff
@@ -0,0 +1,438 @@
+
+        ReadFirstToken:
+            ReadFirstToken(ref reader, first);
+            retVal = true;
+            goto Done;
+        }
+
+        private void ConsumeNextToken(ref BufferReader<byte> reader, byte marker)
+        {
+            reader.Advance(1);
+            CurrentIndex++;
+            if (marker == JsonConstants.ListSeperator)
+            {
+                if (!reader.TryPeek(out byte first))
+                    JsonThrowHelper.ThrowJsonReaderException(ref this);
+
+                if (first <= JsonConstants.Space)
+                {
+                    SkipWhiteSpace(ref reader);
+                    // The next character must be a start of a property name or value.
+                    if (!reader.TryPeek(out first))
+                        JsonThrowHelper.ThrowJsonReaderException(ref this);
+                }
+
+                TokenStartIndex = CurrentIndex;
+                if (_inObject)
+                {
+                    if (first != JsonConstants.Quote)
+                        JsonThrowHelper.ThrowJsonReaderException(ref this);
+
+                    reader.Advance(1);
+                    CurrentIndex++;
+                    TokenStartIndex++;
+                    ConsumePropertyName(ref reader);
+                }
+                else
+                {
+                    ConsumeValue(ref reader, first);
+                }
+            }
+            else if (marker == JsonConstants.CloseBrace)
+            {
+                EndObject();
+            }
+            else if (marker == JsonConstants.CloseBracket)
+            {
+                EndArray();
+            }
+            else
+            {
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+            }
+        }
+
+        private void ConsumeValue(ref BufferReader<byte> reader, byte marker)
+        {
+            TokenType = JsonTokenType.Value;
+
+            if (marker == JsonConstants.Quote)
+            {
+                reader.Advance(1);
+                CurrentIndex++;
+                TokenStartIndex++;
+                ConsumeString(ref reader);
+            }
+            else if (marker == JsonConstants.OpenBrace)
+            {
+                reader.Advance(1);
+                CurrentIndex++;
+                StartObject();
+                ValueType = JsonValueType.Object;
+            }
+            else if (marker == JsonConstants.OpenBracket)
+            {
+                reader.Advance(1);
+                CurrentIndex++;
+                StartArray();
+                ValueType = JsonValueType.Array;
+            }
+            else if ((uint)(marker - '0') <= '9' - '0')
+            {
+                ConsumeNumber(ref reader);
+            }
+            else if (marker == '-')
+            {
+                //TODO: Is this a valid check or do we need to do an Advance to be sure?
+                if (reader.End) JsonThrowHelper.ThrowJsonReaderException(ref this);
```

Comment:
When reviewing recent JsonReader PRs, I have seem lots of code that simply throws when there is not enough data. But we will need to make JsonReader be able to continue after more data arrives. Is that something you thought about? how would we enable starting reasing a ROS that is not complete JSON, getting more data, and continuing parsing.

Reply:
Yes, I plan to add that with a `hasMoreData` flag, which defaults to false that the user passes in. We still need to keep the throw logic for invalid/incomplete json data in the case when the user sets `hasMoreData` as false and we reach the end of data state.